# OPERATORS

## Introduction
___



An operator is a symbol that tells the compiler to perform specific mathematical or logical manipulations. R language is rich in built-in operators and provides following types of operators.

__Types of Operators:__ We have the following types of operators in R programming 

1. Arithmetic Operators
2. Relational Operators
3. Logical Operators
4. Miscellaneous Operators

## 1.0 Arithmetic Operators
___


Following detail shows the arithmetic operators supported by R language. The operators act on each element of the vector.


In [ ]:
# Add two vectors (+)
v1 <- c(10, 20, 30)
v2 <- c(3, 4, 7)
v1 + v2

[1] 13 24 37

In [ ]:
# Subtract two vectors (-)
v1 - v2

[1]  7 16 23

In [ ]:
# Multiply two vectors (*)
v1 * v2

[1]  30  80 210

In [ ]:
# Divide two vectors (/)
v1 / v2

[1] 3.333333 5.000000 4.285714

In [ ]:
# Modulus of two vectors (%%);
# it gives the remainder of the division of the first vector
v1 %% v2

[1] 1 0 2

In [ ]:
# Exponentiation of two vectors (^)
v1 ^ v2

[1] 1.000e+03 1.600e+05 2.187e+10

In [ ]:
# Floor division of two vectors (%/%)
v1 %/% v2

[1] 3 5 4


* The `%/%` operator in R performs floor division, also known as integer division. This operation divides the first vector (v1) by the second vector (v2) and then rounds down to the nearest whole number. The result is a new vector where each element is the result of floor division of the corresponding elements in v1 and v2.  

* For example, if v1 is c(10, 20, 30) and v2 is c(3, 4, 7), the result of v1 %/% v2 would be c(3, 5, 4). This is because 10 divided by 3 is approximately 3.33 (rounded down to 3), 20 divided by 4 is exactly 5, and 30 divided by 7 is approximately 4.28 (rounded down to 4).  

* In summary, this line of code is performing element-wise floor division on two vectors, which can be useful in many data analysis tasks.

## 2.0 Relational Operators
___


Following table shows the relational operators supported by R language. Each element of the first vector is compared with the corresponding element of the second vector. The result of comparison is a Boolean value.


In [ ]:
# Checks if each element of the first vector is greater than
# the corresponding element of the second vector.
v <- c(2, 5.5, 6, 9)
t <- c(8, 2.5, 4, 9)
print(v > t)

[1] FALSE  TRUE  TRUE FALSE


In [ ]:
# Checks if each element of the first vector is less than
# the corresponding element of the second vector.
print(v < t)

[1]  TRUE FALSE FALSE FALSE


In [ ]:
# Checks if each element of the first vector is equal to
# the corresponding element of the second vector.
print(v == t)

[1] FALSE FALSE FALSE  TRUE


In [ ]:
# Checks if each element of the first vector is not equal to
# the corresponding element of the second vector.
print(v != t)

[1]  TRUE  TRUE  TRUE FALSE


In [ ]:
# Checks if each element of the first vector is greater than or equal to
# the corresponding element of the second vector.
print(v >= t)

[1] FALSE  TRUE  TRUE  TRUE


In [ ]:
# Checks if each element of the first vector is less than or equal to
# the corresponding element of the second vector.
print(v <= t)

[1]  TRUE FALSE FALSE  TRUE


## 3.0 Logical Operator
___


Suppose we want to change or compare the results of the comparisons made using relational operators. How would we go about doing that?

 R does this using the AND, OR and NOT operators.


* AND Operator: Represented using an ampersand (&), this operator takes two logical values and returns TRUE only if both values are TRUE themselves.
  
* OR Operator: Denoted using the pike symbol (|), this operator takes two logical values and returns TRUE if just one value is TRUE.

* NOT Operator: Represented using an exclamation mark(!) this operator negates the logical value it’s used on.


### 3.1 "AND"  Operator “&”
___

* The AND operator takes two logical values and returns TRUE only if both values are TRUE themselves. This means that TRUE & TRUE evaluates to TRUE, but that FALSE & TRUE, TRUE & FALSE and FALSE & FALSE evaluates to FALSE.
* Zero is considered FALSE and non-zero numbers are taken as TRUE. Let's see an example for this:

In [ ]:
TRUE & TRUE
TRUE & FALSE
FALSE & FALSE
FALSE & TRUE

[1] TRUE

[1] FALSE

[1] FALSE

[1] FALSE


Instead of using logical values, we can use the results of comparisons. Suppose we have a variable x that is equal to 12. To check if this variable is greater than five but less than 15, we can use x greater than five and x less than 15.

In [ ]:
x <- 12
x > 5 & x < 15

[1] TRUE


The first part, x > 5, will evaluate to TRUE because 12 is greater than five. The second part, x < 15, will also evaluate to TRUE because 12 is also less than 15. So, the result of this expression is TRUE since TRUE & TRUE is TRUE. This makes sense, because 12 lies between five and 15.  

However, if x were 17, the expression x > 5 & x < 15 would simplify to TRUE & FALSE, which would result in the expression being FALSE.


In [ ]:
y <- 17
y > 5 & y < 15

[1] FALSE

In [ ]:
# Example Problem
linkedin <- c(16, 9, 13, 5, 2, 17, 14)
facebook <- c(17, 7, 5, 16, 8, 13, 14)


The linkedin vector represents the views on your LinkedIn profile from the past seven days, and the facebook vector represents the views on your Facebook profile from the past seven days.

Determine when LinkedIn views exceeded 10 and Facebook views failed to reach 10 for a particular day. Use the linkedin and facebook vectors.


In [ ]:
linkedin > 10 & facebook < 10

[1] FALSE FALSE  TRUE FALSE FALSE FALSE FALSE

### 3.2 "OR" Operator “|”
___

The OR operator | works similarly to the AND operator &, but the difference is that only one of the logical values needs to be equal to TRUE for the entire OR operation to evaluate to TRUE.

This means that TRUE | TRUE equals TRUE, but also, TRUE | FALSE and FALSE | TRUE evaluates to TRUE. When both logical are FALSE in an OR operation, FALSE | FALSE, the result is FALSE. Remember, the OR operation is not an exclusive OR operation, so TRUE | TRUE equals TRUE as well.

In [ ]:
TRUE | TRUE
TRUE | FALSE
FALSE | FALSE
FALSE | TRUE

[1] TRUE

[1] TRUE

[1] FALSE

[1] TRUE


With the AND operator, only TRUE & TRUE makes a TRUE, anything else is FALSE. With the OR operator, only FALSE | FALSE makes a FALSE, anything else is TRUE.
Just as in AND operations, we can use comparisons together with the OR operator. Suppose we have a variable y that equals four. To see if this variable is less than five or greater than 15, we can use the following expression:


In [ ]:
y <- 4
y < 5 | y > 15

[1] TRUE


R will first carry out the comparisons, resulting in TRUE | FALSE, which in turn, results in TRUE.
Now, suppose y is 14. The expression y < 5 | y > 15 now evaluates to FALSE | FALSE. Neither one of the comparisons are TRUE, so the result is FALSE.


In [ ]:
y <- 14
y < 5 | y > 15

[1] FALSE

In [ ]:
# Example Problem
linkedin <- c(16, 9, 13, 5, 2, 17, 14)
facebook <- c(17, 7, 5, 16, 8, 13, 14)

# Check if the number of LinkedIn users is greater than 10 or
# the number of Facebook users is less than 10.

linkedin > 10 | facebook < 10

[1]  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE

### 3.3 "Not" Operator “!”
___


The NOT operator, represented by an exclamation mark !, simply negates the logical value it’s used on. That is, !TRUE evaluates to FALSE, while !FALSE evaluates to TRUE.

In [ ]:
!TRUE
!FALSE

[1] FALSE

[1] TRUE

<tt>
Just like the OR and AND operators, we can use the NOT operator in combination with logical operators. This is not always necessary. For example, !(x < 5) is the same as x >= 5.
<tt>

<tt>
- However, there are cases in R where the NOT operator is especially handy. For example, the built-in R function, <code>is.numeric()</code> checks if an R object is a numeric. 

- There is no respective built-in function that checks if it isn’t a numeric. To check, we would have to negate the result <code>(!is.numeric())</code>
  
- So, <code>is.numeric(5)</code> evaluates to TRUE, as five is a numeric. 
  
- If we negate this result using the NOT operator <code>(!is.numeric(5))</code>, we get FALSE. 
  
- If, however, we use <code>is.numeric("hello")</code> we get FALSE. Negating this result <code>(!is.numeric("hello"))</code> gives us TRUE.
<tt>

In [ ]:
is.numeric(5)
is.numeric("hello")

!is.numeric(5)
!is.numeric("hello")

[1] TRUE

[1] FALSE

[1] FALSE

[1] TRUE

In [53]:
x <- 5
y <- 7
!(!(x < 4) & !!!(y > 12))

[1] FALSE


Explanation of !(x < 4)
- !(x < 4) checks if x is less than 4. Since x is 5, this is FALSE. The ! operator then negates this, resulting in TRUE.

_Explanation of !!!(y > 12))_

- y > 12: Since y is seven, and 7 > 12 is not true, this expression evaluates to FALSE.
- !(y > 12): Negating the result from step one, we get !FALSE, or TRUE.
- !!(y > 12): Negating the result from step two, we get !TRUE, or FALSE.
- !!!(y > 12): Negating the result from step three, we get !FALSE, or TRUE.

So, for the inner expression (!(x < 4) & !!!(y > 12)), it evaluates to TRUE & TRUE, which equals TRUE.

The outer NOT operator ! negates this TRUE making !(!(x < 4) & !!!(y > 12)) equal to !TRUE or FALSE.

Thus, the final answer is FALSE

## 4.0 Miscellaneous Operators
___

<tt>
These operators are used to for specific purpose and not general mathematical or logical computation.
<tt>

In [61]:
# colon operator (:) creates a sequence of numbers
# from a starting number to an ending number.
v <- 1:10
print(v)

 [1]  1  2  3  4  5  6  7  8  9 10


In [64]:
# seq() function creates a sequence of numbers
# from a starting number to an ending number.
y <- seq(1, 15)
print(y)

 [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15


In [65]:
# seq() function with by argument
z <- seq(1, 15, by = 2)
print(z)

[1]  1  3  5  7  9 11 13 15


In [66]:
# seq() function with length.out argument
a <- seq(1, 15, length.out = 5)
print(a)

[1]  1.0  4.5  8.0 11.5 15.0


In [69]:
# (%in%) This operator is used to identify if an element belongs to a vector.
# It returns a logical vector
# indicating if there is a match or not for its left operand.
v1 <- 8
v2 <- 12
t <- 1:10
print(v1 %in% t)
print(v2 %in% t)

[1] TRUE
[1] FALSE


In [1]:
# (%*%) This operator is used to perform matrix multiplication.
# It is used to multiply two matrices.
m1 <- matrix(c(3, 2, 5, 6), nrow = 2, ncol = 2)
m2 <- matrix(c(7, 8, 9, 10), nrow = 2, ncol = 2)
print(m1)
print(m2)
print(m1 %*% m2)

     [,1] [,2]
[1,]    3    5
[2,]    2    6
     [,1] [,2]
[1,]    7    9
[2,]    8   10
     [,1] [,2]
[1,]   61   77
[2,]   62   78
